clang/tools/libclang/CCAS.cpp

```diff
@@ -74,6 +138,334 @@ void clang_experimental_cas_Databases_dispose(CXCASDatabases CDBs) {
+  return wrap(new WrappedCASObject{std::move(*Obj), DBs.CAS});
+}
+
+void clang_experimental_cas_loadObjectByString_async(
+    CXCASDatabases CDBs, const char *PrintedID, void *Ctx,
+    void (*Callback)(void *Ctx, CXCASObject, CXError),
+    CXCASCancellationToken *OutToken) {
+  if (OutToken)
+    *OutToken = nullptr;
+  WrappedCASDatabases &DBs = *unwrap(CDBs);
+  ObjectStore &CAS = *DBs.CAS;
+
+  Expected<CASID> Digest = CAS.parseID(PrintedID);
+  if (!Digest)
+    return Callback(Ctx, nullptr, cxerror::create(Digest.takeError()));
+  std::optional<ObjectRef> Ref = CAS.getReference(*Digest);
+  if (!Ref)
+    return Callback(Ctx, nullptr, nullptr);
+
+  /// Asynchronously visits the graph of the object node to ensure it's fully
+  /// materialized.
+  class AsyncObjectLoader
+      : public std::enable_shared_from_this<AsyncObjectLoader> {
+    void *Ctx;
+    void (*Callback)(void *Ctx, CXCASObject, CXError);
+    std::shared_ptr<cas::ObjectStore> CAS;
+
+    llvm::SmallDenseSet<ObjectRef> ObjectsSeen;
+    unsigned NumPending = 0;
+    std::optional<ObjectProxy> RootObj;
+    std::atomic<bool> MissingNode{false};
+    /// The first error that occurred.
+    std::optional<Error> ErrOccurred;
+    std::mutex Mutex;
+
+  public:
+    AsyncObjectLoader(void *Ctx,
+                      void (*Callback)(void *Ctx, CXCASObject, CXError),
+                      std::shared_ptr<cas::ObjectStore> CAS)
+        : Ctx(Ctx), Callback(Callback), CAS(std::move(CAS)) {}
+
+    void visit(ObjectRef Ref, bool IsRootNode) {
+      bool Inserted;
+      {
+        std::lock_guard<std::mutex> Guard(Mutex);
+        Inserted = ObjectsSeen.insert(Ref).second;
+        if (Inserted)
+          ++NumPending;
+      }
+      if (!Inserted) {
+        finishedNode();
+        return;
+      }
+      auto This = shared_from_this();
+      CAS->getProxyAsync(
+          Ref, [This, IsRootNode](Expected<std::optional<ObjectProxy>> Obj) {
+            auto _1 = llvm::make_scope_exit([&]() { This->finishedNode(); });
+            if (!Obj) {
+              This->encounteredError(Obj.takeError());
+              return;
+            }
+            if (!*Obj) {
+              This->MissingNode = true;
+              return;
+            }
+            if (IsRootNode)
+              This->RootObj = *Obj;
+            cantFail((*Obj)->forEachReference([&This](ObjectRef Sub) -> Error {
+              This->visit(Sub, /*IsRootNode*/ false);
+              return Error::success();
+            }));
+          });
+    }
+
+    void finishedNode() {
+      bool FinishedPending;
+      {
+        std::lock_guard<std::mutex> Guard(Mutex);
+        assert(NumPending);
+        --NumPending;
+        FinishedPending = (NumPending == 0);
+      }
+      if (!FinishedPending)
+        return;
+
+      if (ErrOccurred)
+        return Callback(Ctx, nullptr, cxerror::create(std::move(*ErrOccurred)));
+      if (MissingNode)
+        return Callback(Ctx, nullptr, nullptr);
```

Comment:
If the CAS contains the parent node but not the child, we treat it as a completely missing node, not an error? Is that intentional?

Reply:
The rationale is perhaps in the distributed cache side there was garbage collection involved and I find it preferable to treat this as needing to recompute the data than outright fail.